Add checksumtype to the golangci-lint file.

gregory-nisbet opened the pull request:
Some of the types within cedar-go match the sum type idiom (an interface with an unexported method) described in the linter below.

Let's add an explicit check for exhaustiveness for this specific linter.

https://github.com/alecthomas/go-check-sumtype

*Issue #, if available:*

n/a

*Description of changes:*

Add a sum type annotation, add two branches to convertNamespace, both of which are just comments. Let me know if the behavior in those branches should be a little smarter.

gregory-nisbet committed:
Add checksumtype to the golangci-lint file.

Some of the types within cedar-go match the sum type idiom (an interface with an unexported method) described in the linter below.

Let's add an explicit check for exhaustiveness for this specific linter.

https://github.com/alecthomas/go-check-sumtype

*Issue #, if available:*

IDK what issue tracker you use.

*Description of changes:*

Add a sum type annotation, add two branches to convertNamespace, both of which are just comments.
Let me know if the behavior in those branches should be a little smarter.

Signed-off-by: Greg NISBET <gregory.nisbet@gmail.com>

philhassey reviewed (changes requested):
I think there's a typo to fix.

Sorry for the delay on reviewing this, there was some trouble with notifications.

gregory-nisbet committed:
Disable sumtype on Node.

Some nonce types have an embedded Node member and are in the same
package, so they appear to the sum type checker to be real variants
based on the heuristics that that package uses.

There are a few ways to handle this:
1) make the unknownNode and unknownType cases real variants
2) remove the tests that use unknownNode and unknownType
3) rewrite the tests that use unknownNode and unknownType to use a
   different strategy to detect the same defects.

Signed-off-by: Greg NISBET <gregory.nisbet@gmail.com>

gregory-nisbet commented on internal/schema/ast/ast.go:
```diff
@@ -42,7 +42,8 @@ import (
 // is lossy. It loses formatting, such as comments, ordering of fields, etc...
-
+//
+//symtype:decl
```
Yes. However, fixing it opens another can of worms. `unknownType` and `unknownNode`, which are only used in tests, embed a `Node` member and appear to the sum type checker to be real variants.

I couldn't think of a way to proceed that didn't expand the size of the diff enormously, so I just disabled the declaration and added a comment explaining what the error is.

philhassey commented on internal/schema/ast/ast.go:
```diff
@@ -165,6 +177,8 @@ func (t *CommonTypeDecl) End() token.Position {
 // 2. A set type (Set<String>)
 // 3. A path (Namespace::EntityType or String)
+//
+// sumtype:decl
```
I notice this has a space before it whereas the others don't.  It'd be nice if they were consistent.

philhassey reviewed (changes requested):
Just a minor nit.

philhassey commented on internal/schema/ast/ast.go:
```diff
@@ -42,7 +42,18 @@ import (
+//
+// Enabling this causes the following error to be reported:
+//
+// internal/schema/ast/format.go:69:2: exhaustiveness check failed for sum type "Node" (from internal/schema/ast/ast.go:47:6): missing cases for unknownNode, unknownType (gochecksumtype)
```
Then the unnecessary tests can be deleted as well and your PR would prove the exhaustive switches are correct.

gregory-nisbet replied:
Done.

gregory-nisbet committed:
Drop now-redundant tests.

Remove unknownType and unknownNode and the two tests that depend on
them.

Signed-off-by: Greg NISBET <gregory.nisbet@gmail.com>

gregory-nisbet committed:
Add panic for unhandled case in convert_json.go

Signed-off-by: Greg NISBET <gregory.nisbet@gmail.com>

gregory-nisbet commented:
I want to split this PR.

Here's the first piece: https://github.com/cedar-policy/cedar-go/pull/121

patjakdev commented:
Closing PR. The sumtype linter now runs in the [build_and_test](https://github.com/cedar-policy/cedar-go/blob/a9a4b1b1450917d5df2a3c7c17d6567035fd8fcb/.github/workflows/build_and_test.yml#L28) workflow.
